proposals/namespace_editing/README.md

```diff
@@ -0,0 +1,480 @@
+
+For many namespace editing workflows we expect it will be desirable for UsdObjects to persist across namespace edits. This can be particularly useful for GUI workflows that may want to maintain state across undo/redo operations or just want a handle that continues to refer to a renamed or reparented object. For these situations we want to provide the ability for UsdObjects to maintain and update an identity that can: 1) be transferred across rename or reparent (e.g. an existing UsdPrim(</A>) automatically becomes UsdPrim(</B>) if /A is renamed to /B) and 2) be “revived” if the composed object is deleted and at some point later is created again at the same path (e.g. UsdPrim(</A>) becomes invalid when /A is deleted, but if a prim at /A is added again, UsdPrim(</A>) automatically becomes valid again, referring to the new prim).
+
+The maintaining of object identities doesn’t come for free so we do not want to pay the cost for it in the majority of workflows which do not expect to perform namespace edits or do not care about UsdObject identities even in the presence of edits. Therefore we propose that the existing UsdObjects will not maintain their identities themselves; they will remain unchanged. Instead we will introduce a new UsdObjectHandle (plus the equivalent UsdPrimHandle, UsdPropertyHandle, etc.) that will hold a UsdObject and a persistent shared “identity” for the object. The handle will behave as a pointer to the underlying UsdObject and will update the underlying object when needed to refer to the composed object that is now represented by the handle’s identity. UsdObjectHandles will be able to be explicitly constructed from UsdObjects essentially creating identities “on demand”. If need be, we could additionally provide API on UsdStage for getting UsdObjectHandles (parallel to the ones that just return UsdObjects) but for now we expect the explicit construction of the handles to be sufficient.
+
+So for example, if a UsdStage has two prims </A> and its child </A/Child> and you were to get:
+```  
+UsdPrim aPrim = stage.GetPrimAtPath(“/A”)
+UsdPrimHandle childHandle = stageGetPrimHandleAtPath(“/A/Child”)
+```
+
+Both aPrim and childHandle would start valid. If you use the namespace editing API to rename /A to /B, then aPrim will no longer be valid since UsdPrim doesn’t track identity. But childHandle will now hold the prim object that refers to /B/Child as UsdPrimHandle will have its shared identity updated. The same would be true if we had instead deleted /A and then performed some sort of “undo” that recreated /A and /A/Child again. aPrim would still be invalid even though a prim at /A exists again, but childHandle would become valid again and hold a UsdPrim representing the recreated /A/Child.
+
+Risks, Issues and Caveats
+=========================
+
```

Comment:
One other thing I've been thinking about... a UsdNamespaceEditor will need to do alot of work that can be cached to build up the "backwards dependencies" for relationships and connections.  If a DCC wants to provide for fast, incremental namespace edits, do we envision allowing it to retain a single UsdNamespaceEditor per-stage, which listens to general editing notices so that it can update its dependency caches (when rel/con targets change) rather than needing to fully regenerate them for each new batch of edits?

Reply:
I don't necessarily envision the UsdNamespaceEditors being created per-stage as none of the API requires a persistent stage; i.e. the all the functions you would call provide the stage you're editing either explicitly (as UsdStage parameter) or implicitly (via a UsdObject parameter). But I do expect that UsdNamespaceEditor object can be persistent and would hold on to authoring settings and such. So, it's completely feasible in the implementation for it to cache dependencies that are outside of what is already being handled by PcpCaches for any stages that are affected by the edits performed.